libs/language-server/src/lib/ast/expressions/operators/binary-arithmetic-expression.ts

```diff
@@ -0,0 +1,202 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { strict as assert } from 'assert';
+
+import { ValidationContext } from '../../../validation/validation-context';
+import { BinaryExpression } from '../../generated/ast';
+// eslint-disable-next-line import/no-cycle
+import {
+  PropertyValuetype,
+  isNumericType,
+  numericTypes,
+} from '../../model-util';
+import { evaluateExpression } from '../evaluation';
+import {
+  BinaryTypeInferenceFunction,
+  EvaluationFunction,
+  EvaluationStrategy,
+} from '../operator-registry';
+import { generateUnexpectedTypeMessage } from '../type-inference';
+
+export const inferBinaryArithmeticExpressionType: BinaryTypeInferenceFunction =
+  (
+    leftType: PropertyValuetype,
+    rightType: PropertyValuetype,
+    expression: BinaryExpression,
+    context: ValidationContext | undefined,
+  ): PropertyValuetype | undefined => {
+    assert(
+      expression.operator === '+' ||
+        expression.operator === '-' ||
+        expression.operator === '*' ||
+        expression.operator === '/' ||
+        expression.operator === '%',
+    );
+
+    if (!isNumericType(leftType) || !isNumericType(rightType)) {
+      if (!isNumericType(leftType)) {
+        context?.accept(
+          'error',
+          generateUnexpectedTypeMessage(numericTypes, leftType),
+          {
+            node: expression.left,
+          },
+        );
+      }
+      if (!isNumericType(rightType)) {
+        context?.accept(
+          'error',
+          generateUnexpectedTypeMessage(numericTypes, rightType),
+          {
+            node: expression.right,
+          },
+        );
+      }
+      return undefined;
+    }
+    if (
+      leftType === PropertyValuetype.INTEGER &&
+      rightType === PropertyValuetype.INTEGER &&
+      expression.operator !== '/'
+    ) {
+      return PropertyValuetype.INTEGER;
+    }
+    return PropertyValuetype.DECIMAL;
+  };
+
+export const evaluateBinaryMultiplicationExpression: EvaluationFunction<
+  BinaryExpression
+> = (
+  expression: BinaryExpression,
+  strategy: EvaluationStrategy,
+  context: ValidationContext | undefined,
+): boolean | number | string | undefined => {
+  assert(expression.operator === '*');
+  const leftValue = evaluateExpression(expression.left, strategy, context);
+  if (leftValue === undefined && strategy === EvaluationStrategy.LAZY) {
+    return undefined;
+  }
```

Comment:
Ah I see, lazy evaluation introduces the return type `undefined`.

Reply:
The return type `undefined` is not caused by lazy evaluation but rather by arithmetic errors (e.g. division by zero). This code construct skips the evaluation of the right operand if the left one evaluated to `undefined`.

And sure, I can move the evaluation strategy part to the front. I agree that it's probably easier to read that way.